### PR TITLE
fix(gatsby): Use xstate `predictableActionArguments` & update to 4.34

### PR DESCRIPTION
**Pull request by LekoArts**
## Description

Since https://github.com/statelyai/xstate/pull/3289 there's a warning printed to the console now to use this option. Trying this out if tests break 🤷 

Also works around [removal of `meta.state` property](https://github.com/statelyai/xstate/pull/3289/files#diff-bba167ca86c51ba527660a25c956376ee3fcab2781efbe6043e97588e3853c00) mentioned in:

> Please be aware that you might not able to use `state` from the `meta` argument when using this flag.

## Related Issues

[ch54148]

### CODE REVIEW

File: packages/gatsby/src/commands/develop-process.ts [hunk 1]
```diff
@@ -164,6 +164,7 @@ module.exports = async (program: IDevelopArgs): Promise<void> => {
   const service = interpret(machine)
 
   if (program.verbose) {
+    // @ts-ignore - I don't understand this failure
```
[the remark this answers is not in the document]
Comment (LekoArts) [reply packages/gatsby/src/commands/develop-process.ts]: Thank you!

**Comment by pieh** [comment]
@Andarist maybe you can help us out on another one :)

https://github.com/gatsbyjs/gatsby/blob/28cfadeb453ece090ab5919be92599f2b847f5d8/packages/gatsby/src/state-machines/query-running/index.ts#L63-L81

When `predictableActionArguments: true` is used the order of things changes on this state.

Before it would first do `entry` / `assign` and then invoke `calculateDirtyQueries` service, but with changes in this PR, it first invoke `calculateDirtyQueries` service and then does `entry` /  `assign` - is this expected?

The setup for this is done to grab current batch list (`pendingQueryRuns`) and pass it to service, while reseting it globally so if new entry would show up while we are in service it wouldn't get lost (just would need to wait for another for its turn / next batch execution) - is there better setup to achieve something like that?

**Comment by Andarist** [comment]
This is essentially a byproduct of how those are concatenated [here](https://github.com/statelyai/xstate/blob/7d5be188f85bca4d25d79888d4e14bdd6811952e/packages/core/src/StateNode.ts#L1072-L1073). I could reverse this but then it wouldn't be possible to do this:
```ts
entry: send({ type: 'PING' }, { to: 'child' }),
invoke: {
  id: 'child',
  src: childMachine
}
```

We could introduce more custom logic for this case to fix this though as I think it makes sense in this case for `invoke` to have access to the updated `context`.

I need to think about the potential implementation in v4 as this is a little bit tricky - it doesn't help that the upcoming v5 has much better architecture internally, with which this wouldn't be a problem at all 😬

**Comment by pieh** [comment]
> We could introduce more custom logic for this case to fix this though as I think it makes sense in this case for `invoke` to have access to the updated `context`.
> 
> I need to think about the potential implementation in v4 as this is a little bit tricky - it doesn't help that the upcoming v5 has much better architecture internally, with which this wouldn't be a problem at all 😬

Please, do let us know wether we should expect changes in this behavior. Maybe consider temporarily disabling the warning that is printed to console about `predectableActionArguments` until there is clarity on what the behavior of that mode should be? I do think in meantime we will pin `xstate` to previous minor, as our users are seeing those warnings which we are trying to "silence" with this PR

**Comment by Andarist** [comment]
Let me know if you'd like any further assistance with this.

**Comment by LekoArts** [comment]
I initially merged https://github.com/gatsbyjs/gatsby/pull/36393 but that failed some tests and for now I'm fine with it being not optimal but working :D

Appreciate your input and help!

**Review by pieh — APPROVED** [review]

**Pull request merged**